### PR TITLE
[12.0][FIX] account_asset_management: Avoid the error when clicking on "Delete" button if there is a residual value and all the depreciation lines are posted.

### PR DESCRIPTION
**Pull request by victoralmau**
Avoid the error when clicking on `Delete` button if there is a residual value and all the depreciation lines are posted.

Please @pedrobaeza and @joao-p-marques can you review it?

@Tecnativa TT32861

### CODE REVIEW

**Comment by luc-demeyer** [comment]
I think we are fixing the issue in the wrong place with this fix. 

A recompute of the depreciation table in such a case should add the last depreciation line(s) hence I think it goes wrong at this point:

table = asset._compute_depreciation_table()

The _compute_depreciation_table should add unposted line(s) so that the recomputed table's residual value is equal to  the 'salvage value'. With such extra line(s) the logic added by this fix is no longer needed.

**Comment by victoralmau** [comment]
> I think we are fixing the issue in the wrong place with this fix.
> 
> A recompute of the depreciation table in such a case should add the last depreciation line(s) hence I think it goes wrong at this point:
> 
> table = asset._compute_depreciation_table()
> 
> The _compute_depreciation_table should add unposted line(s) so that the recomputed table's residual value is equal to the 'salvage value'. With such extra line(s) the logic added by this fix is no longer needed.



> _compute_depreciation_table

Ok, I have made some change in the `_compute_depreciation_table()` function so that only if necessary that extra line is created.

File: account_asset_management/models/account_asset.py [hunk 1]
```diff
@@ -528,7 +528,7 @@ def group_lines(x, y):
             if posted_lines:
                 last_depreciation_date = last_line.line_date
                 last_date_in_table = table[-1]['lines'][-1]['date']
-                if last_date_in_table <= last_depreciation_date:
+                if last_date_in_table < last_depreciation_date:
```
Comment (pedrobaeza) [inline account_asset_management/models/account_asset.py]: Why this change?

File: account_asset_management/models/account_asset.py [hunk 2]
```diff
@@ -563,6 +563,21 @@ def group_lines(x, y):
                 amount_diff = round(
                     residual_amount_table - residual_amount, digits)
                 if amount_diff:
+                    # auto-create new line only if is needed
+                    total_table_lines = 0
+                    for table_i, entry in enumerate(table):
```
Comment (pedrobaeza) [inline account_asset_management/models/account_asset.py]: `table_i` not used, but why don't you use `mapped` and do everything in one line?

File: account_asset_management/models/account_asset.py [hunk 3]
```diff
@@ -563,6 +563,21 @@ def group_lines(x, y):
...
+                        total_table_lines += len(entry['lines'])
+                    move_check_lines = asset.depreciation_line_ids.filtered(
+                        lambda x: x.move_check
+                    )
+                    if len(move_check_lines) == total_table_lines:
```
Comment (pedrobaeza) [inline account_asset_management/models/account_asset.py:579]: Why this comparison?

**Review by pedrobaeza — COMMENTED** [review]
The code is hard to be understood due to a lot of comparisons. The problem seen outside is easy: the sum of the amounts of the depreciations is different from the amount to depreciate.

**Comment by victoralmau** [comment]
> The code is hard to be understood due to a lot of comparisons. The problem seen outside is easy: the sum of the amounts of the depreciations is different from the amount to depreciate.

I have made some changes and added comments to the code to explain the need for any changes.

**Review by joao-p-marques — APPROVED** [review]
Apart from what was already commented, LGTM :+1:

File: account_asset_management/models/account_asset.py [hunk 4]
```diff
@@ -526,9 +526,22 @@ def group_lines(x, y):
             # recompute in case of deviation
             depreciated_value_posted = depreciated_value = 0.0
             if posted_lines:
+                total_table_lines = sum(
+                    [len(entry[1]["lines"]) for entry in enumerate(table)]
```
Comment (pedrobaeza) [inline account_asset_management/models/account_asset.py]: Why using enumerate? It's not adding value and even complicating you. You don't need it for iterating.

File: account_asset_management/models/account_asset.py [hunk 5]
```diff
@@ -526,9 +526,22 @@ def group_lines(x, y):
...
+                )
+                move_check_lines = asset.depreciation_line_ids.filtered(
+                    lambda x: x.move_check
```
Comment (pedrobaeza) [inline account_asset_management/models/account_asset.py]: You can write simply the name of the field when you want to filter if it's set

File: account_asset_management/models/account_asset.py [hunk 6]
```diff
@@ -526,9 +526,22 @@ def group_lines(x, y):
...
+                )
                 last_depreciation_date = last_line.line_date
                 last_date_in_table = table[-1]['lines'][-1]['date']
-                if last_date_in_table <= last_depreciation_date:
+                # If the number of lines in the table is the same as the depreciation
+                # lines, we will not show an error even if the dates are the same.
+                if (
```
Comment (pedrobaeza) [inline account_asset_management/models/account_asset.py:535]: I don't get why is this needed, as this is checking the integrity. We should keep it or provide a case where it's legit to bypass it for me to analyze it.

File: account_asset_management/models/account_asset.py [hunk 7]
```diff
@@ -563,6 +576,18 @@ def group_lines(x, y):
...
+                    # We will auto-create a new line because the number of lines in
+                    # the tables are the same as the posted depreciations and there
+                    # is still a residual value. Only in this case we will need to
+                    # add a new line to the table with the amount of the difference.
+                    if len(move_check_lines) == total_table_lines:
+                        table[table_i_start]['lines'].append(
+                            table[table_i_start]['lines'][line_i_start-1]
+                        )
+                        line = table[table_i_start]['lines'][line_i_start]
+                        line['date'] = line['date'] + relativedelta(days=1)
```
Comment (pedrobaeza) [inline account_asset_management/models/account_asset.py]: Please put latest date without counting one day more and put 0 in days.

File: account_asset_management/models/account_asset.py (hunk 6 above)
Comment (victoralmau) [reply account_asset_management/models/account_asset.py:535]: I think an example will help explain the need for this change:

**Example A**:
Test asset with a value of 10.
Depreciation table generated:
- 01/01/2021 - 5
- 02/01/2021 - 5

If we regenerate (Compute button) the depreciation table, there is no problem. If we generate the entries for all depreciations, the asset is closed.

**Example B**:
Test asset with a value of 10.
Depreciation table generated:
- 01/01/2021 - 5
- 02/01/2021 - 5

We modify the value of last depreciation to 4.95

If we regenerate the depreciation table, there is no problem. If we generate the entries for all amortizations, the asset is NOT closed because 0.05 will remain pending.

We generate the depreciation table again and the error would already be obtained (the saved date of the last depreciation is equal to the date of the last depreciation that Odoo calculates).

**Review by pedrobaeza — APPROVED** [review]
Is it OK for you, @luc-demeyer ?
